src/home_robot/home_robot/agent/goat_agent/goat_agent_module.py

```diff
@@ -0,0 +1,322 @@
+
+        score_func = self.goal_policy_config.score_function
+        assert score_func in ["confidence_sum", "match_count"]
+
+        if all_matches is not None:
+            self.img_instance_goal_found = False
+            self.goal_inst = None
+            if len(all_matches) > 0:
+                max_scores = []
+                for inst_idx, match_inst in enumerate(all_matches):
+                    inst_view_scores = []
+                    for view_idx, match_view in enumerate(match_inst):
+                        view_score = all_confidences[inst_idx][view_idx][
+                            match_view != -1
+                        ].sum()
+                        inst_view_scores.append(view_score)
+
+                    max_scores.append(max(inst_view_scores))
+                    print(f"Instance {inst_idx+1} score: {max(inst_view_scores)}")
+
+                if max(max_scores) > self.goal_policy_config.score_thresh:
+                    inst_idx = np.argmax(max_scores)
+                    instance_map = local_map[0][
+                        MC.NON_SEM_CHANNELS
+                        + self.num_sem_categories : MC.NON_SEM_CHANNELS
+                        + 2 * self.num_sem_categories,
+                        :,
+                        :,
+                    ]  # TODO: currently assuming img goal instance was an object outside of the vocabulary
+                    inst_map_idx = instance_map == inst_idx + 1
+                    inst_map_idx = torch.argmax(torch.sum(inst_map_idx, axis=(1, 2)))
+                    goal_map_temp = (instance_map[inst_map_idx] == inst_idx + 1).to(
+                        torch.float
+                    )
+
+                    if goal_map_temp.any():
+                        self.img_instance_goal_found = True
+                        self.goal_inst = inst_idx + 1
+                        goal_map = goal_map_temp
+                        print(f"{self.goal_inst} will be the goal")
+                    else:
+                        print("Instance was seen, but not present in local map.")
+                else:
+                    print("Goal image does not match any instance.")
+                    # TODO: dont stop at the first instance, but rather find the best one
+
+        if self.goal_inst is not None and self.img_instance_goal_found is True:
+            found_goal[0] = True
+
+            instance_map = local_map[0][
+                MC.NON_SEM_CHANNELS
+                + self.num_sem_categories : MC.NON_SEM_CHANNELS
+                + 2 * self.num_sem_categories,
+                :,
+                :,
+            ]
+            inst_map_idx = instance_map == self.goal_inst
+            inst_map_idx = torch.argmax(torch.sum(inst_map_idx, axis=(1, 2)))
+            goal_map = (instance_map[inst_map_idx] == self.goal_inst).to(torch.float)
+
+            # goal_map = (local_map[0][22:38][0] == self.goal_inst).to(torch.float)
+        else:
```

Comment:
This `else` condition when `all_matches is None` to match against the current frame is broken.

Just like we match instances in memory when `all_matches is not None`, we should be matching **instances in the current frame** when `all_matches is None`. Matching the goal with an instance gives us a location to navigate to.

We should 
* have confidence scores per detection (bounding box) in the current frame
* be able to associate detections in the current frame with instances in the instance memory — this allows us to select a location as the goal and navigate to it